### PR TITLE
Tensor .cuda() very slow with specific array sizes

### PR DESCRIPTION
**Pull request by tolleybot**
#### **Summary**
This PR addresses the performance issue identified in [#111570](https://github.com/pytorch/pytorch/issues/111570), where non-contiguous tensors took significantly longer to transfer from CPU to GPU. Through detailed tracing of the call flow, we identified that PyTorch was creating temporary contiguous buffers for non-contiguous tensor transfers, which introduced unnecessary overhead.

#### **Tracing the Issue**
To pinpoint the cause of the slowdown, we followed the call flow from Python’s `tensor.cuda()` method through PyTorch’s backend, ultimately identifying `copy_kernel_cuda` as the key function responsible for CPU-to-GPU tensor transfers. Here’s a summary of the tracing process:

1. **Python Call: `tensor.cuda()`**
   - Starting from Python, the `cuda()` method initiates the tensor transfer to the GPU.

2. **`TensorBody.h: cuda()`**
   - The `cuda()` method calls `to()`, specifying the target device as CUDA.

3. **`Tensor.cpp: TensorBase::to()`**
   - The `to()` function prepares device and data type options before invoking `_ops::to_dtype_layout::call()`.

4. **Operator Call: `_ops::to_dtype_layout::call()`**
   - This operator dispatches the request to the backend-specific function responsible for managing the transfer.

5. **`Copy.cpp: copy_()`**
   - The `copy_()` function performs preliminary checks (e.g., zero-tensor immutability) and proceeds to call `copy_impl()`.

6. **`Copy.cpp: copy_impl()`**
   - This function sets up a tensor iterator and dispatches the copy operation to the appropriate backend through `copy_stub`.

7. **Dispatch to CUDA: `copy_stub`**
   - The dispatch mechanism routes the call to the CUDA-specific function, `copy_kernel_cuda`.

8. **`Copy.cu: copy_kernel_cuda()`**
   - Here, we identified that PyTorch was creating temporary contiguous buffers for 1D and 2D non-contiguous tensors, which slowed down the copy process. This behavior is managed by the `copy_requires_temporaries()` function.

#### **Solution**
To address this, we modified `copy_kernel_cuda` to handle non-contiguous 1D and 2D tensors directly by using `cudaMemcpy2DAsync`, which allows efficient, stride-aware memory transfers without temporary buffers. Here’s why this approach improves performance:

- **Efficiency of `cudaMemcpy2DAsync`**: This CUDA function is optimized for pitched (stride-based) memory transfers, allowing it to handle non-contiguous data layouts effectively by specifying memory strides for source and destination tensors.
- **Reduction of Overhead**: By directly copying non-contiguous tensors without intermediate buffers, we eliminate extra memory allocation and achieve faster CPU-to-GPU transfers.
- **Asynchronous Execution**: `cudaMemcpy2DAsync` enables asynchronous transfer on the CUDA stream, further improving performance by taking advantage of CUDA's optimized memory handling for non-contiguous layouts.

#### **Performance Results**

In my testing, I created tensors of size `327680 x 2000` and used slices for transfer performance measurements. The tests show that the average time for transferring a non-contiguous slice (e.g., rows 10,000 to 50,000) from CPU to GPU now closely matches the contiguous case. This improvement indicates that the updated implementation effectively addresses the performance discrepancy. Below are the measured times and validation checks:

```plaintext
Average time for contiguous slice (rows 10,000-50,000): 66 ms
Average time for non-contiguous slice (rows 10,000-50,000): 66 ms

Validation of contiguous and non-contiguous tensor copies:
✅ PASS: Tensor shapes match.
✅ PASS: Tensor contiguity matches.
✅ PASS: Tensor contents match.
✅ PASS: Tensor data types match.

✅ Success: Both contiguous and non-contiguous tensors were copied correctly to the GPU.
```

#### **Conclusion**
This PR resolves the identified performance issue by eliminating the need for temporary buffers in non-contiguous 1D and 2D tensor transfers, ensuring faster and more efficient copies from CPU to GPU. Future optimizations could further enhance performance for higher-dimensional non-contiguous tensors.

### CODE REVIEW

**Comment by cyyever** [comment]
```
Average time for non-contiguous slice (rows 10,000-50,000): 66 ms
```
I'm curious to know the time before this optimization.

File: aten/src/ATen/native/cuda/Copy.cu [hunk 1]
```diff
@@ -380,8 +386,56 @@ static void copy_kernel_cuda(TensorIterator& iter, bool non_blocking) {
     // TODO: warn on the return value.
     CachingHostAllocator_recordEvent(ptr, ctx, stream);
 
-  } else {
-    at::cuda::memcpy_and_sync(dst, src, nbytes, kind, stream);
+  } else { 
+    if (iter.ndim() <= 2 && !iter.is_contiguous()) {
```
Comment (eqy) [inline aten/src/ATen/native/cuda/Copy.cu:390]: nit maybe move this logic to a shared function with the check above to ensure it stays in sync

Comment (tolleybot) [reply aten/src/ATen/native/cuda/Copy.cu:467]: Thans for the feedback. I’ve refactored the code to move the 1D and 2D non-contiguous copy logic into a new shared function, copy_non_contiguous_2d

File: aten/src/ATen/native/cuda/Copy.cu [hunk 2]
```diff
@@ -291,6 +291,83 @@ static bool maybe_enable_p2p_access(Device dst_device, Device src_device) {
...
+      width_in_bytes = dim0 * element_size;
+      height = dim1;
+  } else {
+      src_pitch = stride0 * element_size;
+      dst_pitch = dst_stride0 * element_size;
+      width_in_bytes = dim1 * element_size;
+      height = dim0;
+  }
+
+  TORCH_CHECK(src_pitch >= width_in_bytes, "Source pitch must be >= width_in_bytes");
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Copy.cu:356]: you should not error out in this case, instead you should check that those conditions are satisfied before taking the fast path. This is valid data to copy

File: aten/src/ATen/native/cuda/Copy.cu [hunk 3]
```diff
@@ -380,8 +463,12 @@ static void copy_kernel_cuda(TensorIterator& iter, bool non_blocking) {
     // TODO: warn on the return value.
     CachingHostAllocator_recordEvent(ptr, ctx, stream);
 
-  } else {
-    at::cuda::memcpy_and_sync(dst, src, nbytes, kind, stream);
+  } else {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Copy.cu:466]: you are fixing only the blocking case, whereas usually non-blocking are most-performance sensitive. Please fix non-blocking case also.

**Review by ngimel — COMMENTED** [review]
Please make sure that there are tests that are exercising this code path, or add those tests if necessary.

File: aten/src/ATen/native/cuda/Copy.cu (hunk 2 above)
Comment (tolleybot) [reply aten/src/ATen/native/cuda/Copy.cu:356]: Your correct.  I reworked to the code to handle this case.

File: aten/src/ATen/native/cuda/Copy.cu (hunk 3 above)
Comment (tolleybot) [reply aten/src/ATen/native/cuda/Copy.cu:466]: I have updated the code to handle the non-blocking case.

File: aten/src/ATen/native/cuda/Copy.cu [hunk 4]
```diff
@@ -265,6 +265,12 @@ static bool copy_requires_temporaries(TensorIterator& iter, bool p2p_enabled) {
   }
 
   bool same_dtype = iter.dtype(0) == iter.dtype(1);
+
+  // Check if the tensor is 1D or 2D and non-contiguous
+  if (iter.ndim() <= 2 && !iter.is_contiguous()) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Copy.cu:270]: you also should check for `same_dtype` here? Please add corresponding test case.

Comment (tolleybot) [reply aten/src/ATen/native/cuda/Copy.cu:270]: I don't think the test is needed as its handled by the next if condition.  Im working on a couple of test cases now.

Comment (ngimel) [reply aten/src/ATen/native/cuda/Copy.cu:270]: temporaries are requires if there's a non-contiguous iterator of dim2 (the condition above is satisfied) but dtypes are different. You would already return `false` in this case.

File: aten/src/ATen/native/cuda/Copy.cu [hunk 5]
```diff
@@ -359,6 +454,13 @@ static void copy_kernel_cuda(TensorIterator& iter, bool non_blocking) {
   int64_t nbytes = iter.numel() * iter.element_size(0);
   CUDAStream stream = getCurrentCUDAStream();
 
+  // Try optimized 1D/2D non-contiguous path first for both blocking and non-blocking
+  if (iter.ndim() <= 2 && !iter.is_contiguous()) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Copy.cu:458]: I would really prefer not to send 1d case here, that's a lot of extra work to set sizes and pitch, and unclear if cudaMemcpy2DAsync doesn't have perf regressions compared to cudaMemcpyAsync

**Review by ngimel — COMMENTED** [review]
There are still failing tests

File: aten/src/ATen/native/cuda/Copy.cu (hunk 4 above)
Comment (tolleybot) [reply aten/src/ATen/native/cuda/Copy.cu:288]: understood.  
I have added a couple of test cases

Comment (tolleybot) [reply aten/src/ATen/native/cuda/Copy.cu]: I updated the logic to ensure temporaries are created for non-contiguous 1D/2D tensor copies with differing dtypes

File: test/test_cuda.py [hunk 6]
```diff
@@ -3324,6 +3325,45 @@ def test_gds_fails_in_ci(self):
...
+    
+    def test_1d_copy(self):
+        # Contiguous 1D tensor
+        x = torch.ones(10000000, dtype=torch.uint8)
+        self._test_copy(x, non_blocking=True)
+        self._test_copy(x, non_blocking=False)
+
+        # Non-contiguous 1D tensor using a Fortran-ordered NumPy array
+        fortran_arr = np.asfortranarray(np.ones(10000000, dtype=np.uint8))
+        x = torch.from_numpy(fortran_arr[::2])  # Slice to make it non-contiguous
```
Comment (ngimel) [inline test/test_cuda.py:3347]:
```suggestion
x=torch.ones(1000000, dtype=torch.uint8)[::2]
```

File: test/test_cuda.py [hunk 7]
```diff
@@ -3324,6 +3325,45 @@ def test_gds_fails_in_ci(self):
...
+    def test_2d_copy(self):
+        rows, cols = 1000, 1000
+        # Contiguous 2D tensor
+        x = torch.ones((rows, cols), dtype=torch.float32)
+        self._test_copy(x, non_blocking=True)
+        self._test_copy(x, non_blocking=False)
+
+        # Non-contiguous 2D tensor using Fortran-ordered array
+        fortran_arr = np.asfortranarray(np.random.randn(rows, cols).astype(np.float32))
+        x = torch.from_numpy(fortran_arr)
```
Comment (ngimel) [inline test/test_cuda.py:3361]:
```suggestion
        x = torch.rand(rows, cols).t()
```

File: aten/src/ATen/native/cuda/Copy.cu [hunk 8]
```diff
@@ -377,6 +472,13 @@ static void copy_kernel_cuda(TensorIterator& iter, bool non_blocking) {
   int64_t nbytes = iter.numel() * iter.element_size(0);
   CUDAStream stream = getCurrentCUDAStream();
 
+  // Try optimized 1D/2D non-contiguous path first for both blocking and non-blocking
+  if (iter.ndim() <= 2 && !iter.is_contiguous()) {       
+    if (copy_non_contiguous_2d(dst, src, iter, kind, stream, non_blocking)) {
+      return;
```
[the remark this answers is not in the document]
Comment (tolleybot) [reply aten/src/ATen/native/cuda/Copy.cu:478]: I added the boolean return to copy_non_contigous_2d because previousl it was asserting if this condition was met (src_pitch < width_in_bytes || dst_pitch < width_in_bytes).  I changed this due to a previous comment.  I set it so if the condition was not met to have it return and have the original code be called.  Ill take a look and see if there is a better way to handle this.

File: test/test_cuda.py [hunk 9]
```diff
@@ -3324,6 +3325,43 @@ def test_gds_fails_in_ci(self):
...
+        rows, cols = 1000, 1000
+        # Contiguous 2D tensor
+        x = torch.ones((rows, cols), dtype=torch.float32)
+        self._test_copy(x, non_blocking=True)
+        self._test_copy(x, non_blocking=False)
+
+        # Non-contiguous 2D tensor using Fortran-ordered array
+        x = torch.rand(rows, cols).t()
+        
+        self.assertFalse(x.is_contiguous())
```
[the remark this answers is not in the document]
Comment (tolleybot) [reply test/test_cuda.py:3356]: Thanks for that code bit.  I've updated the tests and validated they do actuallyu call the 2d copy code

File: aten/src/ATen/native/cuda/Copy.cu (hunk 8 above)
[the remark this answers is not in the document]
Comment (tolleybot) [reply aten/src/ATen/native/cuda/Copy.cu:512]: 1. Updated copy_requires_temporaries: Add stride and pitch checks to determine if 1D or 2D non-contiguous tensors can be directly copied without temporaries. Returns false when direct copy conditions are met, avoiding unnecessary memory allocations
  2. Updated copy_kernel_cuda logic
  3. Handle conj and neg adjustments consistently for copied data
  4. Updated test_cuda.py with some formatting and removing numpy import

File: aten/src/ATen/native/cuda/Copy.cu [hunk 10]
```diff
@@ -303,6 +337,95 @@ static bool maybe_enable_p2p_access(Device dst_device, Device src_device) {
...
+  } else {
+      src_pitch = stride0 * element_size;
+      dst_pitch = dst_stride0 * element_size;
+      width_in_bytes = dim1 * element_size;
+      height = dim0;
+  }
+
+  // if conditions not met, fallback needed
+  if (src_pitch < width_in_bytes || dst_pitch < width_in_bytes) {
+    return false;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Copy.cu:411]: you should never return `false` because this case is not handled correctly. You should assert (TORCH_INTERNAL_ASSERT) here that conditions are met

**Review by ngimel — COMMENTED** [review]
There are still test failures.

File: aten/src/ATen/native/cuda/Copy.cu (hunk 10 above)
Comment (tolleybot) [reply aten/src/ATen/native/cuda/Copy.cu:411]: Thanks for the catch, @ngimel. I’ll replace return false with TORCH_INTERNAL_ASSERT to enforce the conditions as you suggested. I’ll also review the test cases today to identify any issues

**Comment by ngimel** [comment]
Do you know why the code errors on ROCm? Do the same tests pass on cuda?

**Comment by ngimel** [comment]
@tolleybot please fix failing tests
cc @jeffdaily, in previous version of this PR rocm tests were failing https://github.com/pytorch/pytorch/pull/138964#issuecomment-2529101536 (now cudaMemcpy2D optimization is limited to nvidia only), can someone on AMD side take a look?

**Comment by jeffdaily** [comment]
> @tolleybot please fix failing tests cc @jeffdaily, in previous version of this PR rocm tests were failing [#138964 (comment)](https://github.com/pytorch/pytorch/pull/138964#issuecomment-2529101536) (now cudaMemcpy2D optimization is limited to nvidia only), can someone on AMD side take a look?

Do you have a link to the failed CI run?  

I built the PR branch on an MI200 inside a docker container that matches latest upstream CI docker image.  I removed the #ifdef's before building so that ROCm would take the same code path.  The new copy tests are both passing for me.

**Comment by ngimel** [comment]
@jeffdaily the failing tests were not the new copy tests, it was some flexattention tests I think? But unfortunately there are no links now. @tolleybot do you remember what the failing tests were?

**Comment by jeffdaily** [comment]
Lint errors seem legit?

**Comment by ngimel** [comment]
@tolleybot can you fix the lint errors please?

**Comment by jeffdaily** [comment]
I took care of the merge conflict.  I couldn't repro the lintrunner failure locally either.

**Review by jeffdaily — DISMISSED** [review]

**Comment by ngimel** [comment]
Lint still complains on the Copy.cu file (btw recently I also got different results from local lintrunner and lintrunner on PR)

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your PR but some slow test start failing after this lands' -c nosignal

test_ops.py::TestCommonCUDA::test_compare_cpu_linalg_lstsq_cuda_float32 [GH job link](https://github.com/pytorch/pytorch/actions/runs/13077344273/job/36500419828) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/98f87edd233ea69cee5f3e73e9eb4b5ab77aa744)

**Comment by ngimel** [comment]
whew that was a near miss in a pretty popular op, thanks for the revert @huydhn

File: aten/src/ATen/native/cuda/Copy.cu [hunk 11]
```diff
@@ -283,6 +319,50 @@ static bool copy_requires_temporaries(TensorIterator& iter, bool p2p_enabled) {
...
+  bool is_complex = at::isComplexType(iter.dtype(1));
+  bool broadcast =  broadcast_required(iter.tensor(0), iter.tensor(1));
+
+  if (iter.ndim() == 2 && same_dtype && !broadcast && !is_complex) {
+    const auto& src_tensor = iter.tensor(1);
+    const auto& dst_tensor = iter.tensor(0);
+
+    size_t element_size = src_tensor.element_size();
+    int64_t dim0 = src_tensor.size(0);
+    int64_t dim1 = iter.ndim() == 1 ? 1 : src_tensor.size(1);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Copy.cu:332]: this is wrong, iter.ndim() and tensor dimensions don't necessarily correspond 1-to-1, because iterator constructor does dimension permutation and coalescing to improve memory accesses. In particular, in the failing test the tensors are 
```
sizes0  [3, 3, 3]
strides0 [12, 1, 4]
sizes1 [3, 3, 3]
strides1 [9, 1, 3]
```

File: aten/src/ATen/native/cuda/Copy.cu [hunk 12]
```diff
@@ -283,6 +319,50 @@ static bool copy_requires_temporaries(TensorIterator& iter, bool p2p_enabled) {
...
+#if defined(__CUDACC__) && !defined(__HIP__)
+  bool is_complex = at::isComplexType(iter.dtype(1));
+  bool broadcast =  broadcast_required(iter.tensor(0), iter.tensor(1));
+
+  if (iter.ndim() == 2 && same_dtype && !broadcast && !is_complex) {
+    const auto& src_tensor = iter.tensor(1);
+    const auto& dst_tensor = iter.tensor(0);
+
+    size_t element_size = src_tensor.element_size();
+    int64_t dim0 = src_tensor.size(0);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/Copy.cu:331]: the checks should be done based on `iter` strides, not on original tensors, because original tensor dimensions are coalesced and reordered in iter, see #146256

Comment (tolleybot) [reply aten/src/ATen/native/cuda/Copy.cu:331]: Your correct.  I'll rework that.  I'm also looking into the test failure.

**Comment by ngimel** [comment]
@tolleybot thanks for the good optimization idea, https://github.com/pytorch/pytorch/pull/146256 is merged now.
